app/controllers/sessions_controller.rb

```diff
@@ -9,7 +9,43 @@ def create
       session[:user_id] = user.id
       redirect_to root_path, notice: "Welcome back, #{current_user.first_name}"
     else
-      redirect_to login_path
+      process_form_login
     end
   end
+
+  def destroy
+    @current_user = nil
+    session.delete :user_id
+    redirect_to root_path, notice: "Logged out Successfully!"
+  end
+
+  private
+
+  def process_form_login
+    user = RegularUser.find_by_email(params[:email])
+    return if user_is_nil?(user)
+    if user.authenticate(params[:password])
+      login_successful user
+    else
+      password_invalid
+    end
+  end
+
+  def user_is_nil?(user)
+    return if user
+    flash["errors"] = ["Log in failed. Email not valid."]
+    redirect_to login_path
+
+    true
+  end
+
+  def login_successful(user)
+    session[:user_id] = user.id
+    redirect_to root_path, notice: "Welcome back, #{current_user.first_name}"
+  end
+
+  def password_invalid
+    flash["errors"] = ["Password incorrect. Please try again."]
+    redirect_to login_path
+  end
 end
```

Comment:
The target is easy maintainability, and separation of concerns. That's why each method is responsible for just one function. About the login error shown, I feel it enhances user experience. Websites like Facebook, Google, Jumia let you know what aspects are responsible for login failure.